submissions/Vlada-khom/tiny-js-world/index.js

```diff
@@ -0,0 +1,72 @@
+/* Refer to https://github.com/OleksiyRudenko/a-tiny-JS-world for the task details
+   Complete the below for code reviewers' convenience:
+
+   Code repository: _put repo URL here_
+   Web app: _put project's github pages URL here_
+   */
+
+// ======== OBJECTS DEFINITIONS ========
+let man = {
+  entity: "person",
+  name: "John",
+  sex: "male",
+  hands: 2,
+  legs: 2,
+  say: "Hello world!",
+  friends: "Jenny, Rex",
+};
+
+let woman = {
+  entity: "person",
+  name: "Jenny",
+  sex: "female",
+  hands: 2,
+  legs: 2,
+  say: "Hello world! I'm Jenny.",
+  friends: "Tom",
+};
+
+let dog = {
+  entity: "dog",
+  name: "Rex",
+  sex: "male",
+  hands: 0,
+  legs: 4,
+  say: "Gav!",
+  friends: "John, Jenny",
+};
+
+let cat = {
+  entity: "cat",
+  name: "Tom",
+  sex: "male",
+  hands: 0,
+  legs: 4,
+  say: "Meyow!",
+  friends: "Jenny",
+};
+
+// ======== OUTPUT ========
+/* Use print(message) for output.
+   Default tag for message is <pre>. Use print(message,'div') to change containing element tag.
+
+   Message can contain HTML markup. You may also tweak index.html and/or styles.css.
+   However, please, REFRAIN from improving visuals at least until your code is reviewed
+   so code reviewers might focus on a single file that is index.js.
+   */
+
+   const inhabitants = [man, woman, cat, dog];
+   const properties = ['entity', 'name', 'sex', 'hands', 'legs', 'say', 'friends'];
+   let characteristicsOfInhabitants = inhabitants.forEach((inhabitant)=> {
+     print(properties.map(property => [inhabitant[property]]).join('; '))
```

Comment:
What is the result of this `.map`? Observe in console
Why do we need to wrap `inhabitant[property]` in square brackets?

Reply:
Clear, I wrap up every property in array. Result map its array [person, John, male, 2, 2 and etc],[person, Jenny,...]. forEach iterate array but return nothing... at the end we must get [],[],[],[].join(transform to string)...
okay, I have questions, if methods Object.values, keys and entries doesn't guarantee ordered. Does Object.fromEntries again not guarantee ordered? If yes, when better use these methods?

Reply:
> I wrap up every property in array
... which we do not actually need. `['entity', 'name']` should turn into `['human', 'John']`, not `[['human'], ['John']]`. Luckily, `[['a']].join(';')` turns into `'a'` due to implicit array extrapolation.
Check what is the result of `[['human', 'John'], ['male', 2, ["a","b"]]].join('; ')` to see the effect of nested arrays. At times this may be what you actually want, but often results in bugs and may trick the developer when they try to explore/debug data structures.

Guess, now you know enough to polish the code.

> if methods Object.values, keys and entries doesn't guarantee order. Does Object.fromEntries again not guarantee order? If yes, when better use these methods?

It doesn't either. They are useful when the order doesn't matter.
A use case: you need a list of some object properties that are also a part of some reference list.
```
const obj = { propF: 11,  propX: 1, propB: 7, propA: 121 };
const requiredProps = ['propA', 'propD', 'propX'].sort();
const intersectionOfPropNames = Object.keys(obj).filter(key => requiredProps.includes(key)); // ['propX', 'propA']
// next we could probably tell if all required props are present
intersectionOfPropNames.length === requiredProps.length // naive
intersectionOfPropNames.sort().join(';') === requiredProps.join(';') // naive 2
JSON.from(intersectionOfPropNames.sort()) === JSON.from(requiredProps) // advanced
requiredProps.reduce((noMisses, requiredProp) => noMisses && intersectionOfPropNames.includes(requiredProp), true) // pro
```

Use case: Transform all values in an object where properties are not predefined.
```
const envVariables = {
  apiKey: 'API_Key',
  ghActor: 'Git_User',
};

function objectValuesToUppercas(obj) {
  return Object.fromEntries(
    Object.entries(obj).map(([key, value]) => [key, value.toUpperCase()])
  );
}

objectValuesToUppercas(envVariables);
```

Reply:
Clear, thank you) What is mean JSON.from in first example, I didn't find?

Reply:
I meant `JSON.stringify`